pkg/cmd/experimental/syncgroups/sync-groups.go

```diff
@@ -0,0 +1,188 @@
+	knownScopes := util.NewStringSet(string(GroupSyncScopeAll), string(GroupSyncScopeWhitelist))
+	return knownScopes.Has(string(scope))
+}
+
+// GroupSyncSource determines the source of the groups to be synced
+type GroupSyncSource string
+
+const (
+	// GroupSyncSourceLDAP determines that the groups to be synced are determined from an LDAP record
+	GroupSyncSourceLDAP GroupSyncSource = "LDAP"
+	// GroupSyncSourceOpenShift determines that the groups to be synced are determined from OpenShift records
+	GroupSyncSourceOpenShift GroupSyncSource = "OpenShift"
+)
+
+func ValidateSource(source GroupSyncSource) bool {
+	knownSources := util.NewStringSet(string(GroupSyncSourceLDAP), string(GroupSyncSourceOpenShift))
+	return knownSources.Has(string(source))
+}
+
+type SyncGroupsOptions struct {
+	// Source determines the source of the list of groups to sync
+	Source GroupSyncSource
+
+	// Scope determines the scope of the group sync
+	Scope GroupSyncScope
+
+	// ConfigSource is the path to the sync config
+	ConfigSource string
+
+	// WhitelistContents are the contents of the whitelist - either names of OpenShift groups given by
+	// an administrator on the command line invocation of the sync-groups command or LDAP group UIDs
+	// read from a whitelist file
+	WhitelistContents []string
+
+	// GroupsInterface is the interface used to interact with OpenShift Group objects
+	GroupInterface osclient.GroupInterface
+
+	// Stderr is the writer to write warnings and errors to
+	Stderr io.Writer
+
+	// Out is the writer to write output to
+	Out io.Writer
+
+	// Quiet controls whether warnings are output
+	Quiet bool
+}
+
+func NewSyncGroupsOptions() *SyncGroupsOptions {
+	return &SyncGroupsOptions{
+		Stderr:            os.Stderr,
+		WhitelistContents: []string{},
+	}
+}
+
+func NewCmdSyncGroups(name, fullName string, f *clientcmd.Factory, out io.Writer) *cobra.Command {
+	options := NewSyncGroupsOptions()
+	options.Out = out
+
+	cmd := &cobra.Command{
+		Use:     fmt.Sprintf("%s SOURCE SCOPE [WHITELIST] --sync-config=CONFIG-SOURCE", name),
+		Short:   "Sync OpenShift Groups with records from an external provider.",
+		Long:    syncGroupsLong,
+		Example: fmt.Sprintf(syncGroupsExamples, fullName),
+		Run: func(c *cobra.Command, args []string) {
+			if err := options.Complete(args, f); err != nil {
+				cmdutil.CheckErr(cmdutil.UsageError(c, err.Error()))
+			}
+
+			if err := options.Validate(); err != nil {
+				cmdutil.CheckErr(cmdutil.UsageError(c, err.Error()))
+			}
+
+			syncer, err := options.CreateSyncer()
+			cmdutil.CheckErr(err)
+
+			errs := syncer.Sync()
+			cmdutil.CheckErr(kerrs.NewAggregate(errs))
+		},
+	}
+
+	cmd.Flags().BoolVarP(&options.Quiet, "quiet", "q", options.Quiet, "Suppress warnings")
+	cmd.Flags().StringVar(&options.ConfigSource, "sync-config", "", "The path to the sync config")
+	cmdutil.AddPrinterFlags(cmd)
+
+	return cmd
+}
+
+func (o *SyncGroupsOptions) Complete(args []string, f *clientcmd.Factory) error {
```

Comment:
Have this method do all the config parsing and testing required to build exactly the correct values for the `LDAPGroupSyncer`.  This is the **only** spot in all the code that inspects all the information.  Everything else takes the exact minimum required.